crates/core/src/expr/expression.rs

```diff
@@ -1,105 +1,220 @@
-			l: Value::Null,
-			o: Operator::default(),
-			r: Value::Null,
-		}
-	}
-}
+	Closure(Box<Closure>),
 
-impl Expression {
-	/// Create a new binary expression
-	pub fn new(l: Value, o: Operator, r: Value) -> Self {
-		Self::Binary {
-			l,
-			o,
-			r,
-		}
-	}
+	Break,
+	Continue,
+	Return(Box<OutputStatement>),
+	Throw(Box<Expr>),
+
+	IfElse(Box<IfelseStatement>),
+	Select(Box<SelectStatement>),
+	Create(Box<CreateStatement>),
+	Update(Box<UpdateStatement>),
+	Upsert(Box<UpsertStatement>),
+	Delete(Box<DeleteStatement>),
+	Relate(Box<RelateStatement>),
+	Insert(Box<InsertStatement>),
+	Define(Box<DefineStatement>),
+	Remove(Box<RemoveStatement>),
+	Rebuild(Box<RebuildStatement>),
+	Alter(Box<AlterStatement>),
+	Info(Box<InfoStatement>),
+	Foreach(Box<ForeachStatement>),
+	Let(Box<SetStatement>),
+	Sleep(Box<SleepStatement>),
 }
 
-impl Expression {
-	pub(crate) fn writeable(&self) -> bool {
+impl Expr {
+	/// Check if this expression does only reads.
+	pub(crate) fn read_only(&self) -> bool {
 		match self {
-			Self::Unary {
-				v,
+			Expr::Literal(_)
+			| Expr::Param(_)
+			| Expr::Table(_)
+			| Expr::Mock(_)
+			| Expr::Constant(_)
+			| Expr::Break
+			| Expr::Continue
+			| Expr::Info(_)
+			| Expr::Sleep(_) => true,
+
+			Expr::Idiom(x) => x.read_only(),
+			Expr::Block(block) => block.read_only(),
+			Expr::Prefix {
+				expr,
+				..
+			}
+			| Expr::Postfix {
+				expr,
 				..
-			} => v.writeable(),
-			Self::Binary {
-				l,
-				r,
+			} => expr.read_only(),
+			Expr::Binary {
+				left,
+				right,
 				..
-			} => l.writeable() || r.writeable(),
+			} => left.read_only() && right.read_only(),
+			Expr::FunctionCall(function) => function.read_only(),
+			Expr::Return(expr) => expr.read_only(),
+			Expr::Throw(expr) => expr.read_only(),
+			Expr::IfElse(s) => s.read_only(),
+			Expr::Select(s) => s.read_only(),
+			Expr::Let(s) => s.read_only(),
+			Expr::Foreach(s) => s.read_only(),
+			Expr::Closure(_) => true,
+			Expr::Create(_)
+			| Expr::Update(_)
+			| Expr::Delete(_)
+			| Expr::Relate(_)
+			| Expr::Insert(_)
+			| Expr::Define(_)
+			| Expr::Remove(_)
+			| Expr::Rebuild(_)
+			| Expr::Upsert(_)
+			| Expr::Alter(_) => false,
 		}
 	}
 
-	/// Checks whether all expression parts are static values
+	/// Checks if a expression is 'pure' i.e. does not rely on the environment.
 	pub(crate) fn is_static(&self) -> bool {
 		match self {
-			Self::Unary {
-				v,
+			Expr::Literal(literal) => literal.is_static(),
+			Expr::Constant(_) => true,
+			Expr::Prefix {
+				expr,
 				..
-			} => v.is_static(),
-			Self::Binary {
-				l,
-				r,
+			} => expr.is_static(),
+			Expr::Postfix {
+				expr,
 				..
-			} => l.is_static() && r.is_static(),
+			} => expr.is_static(),
+			Expr::Binary {
+				left,
+				right,
+				..
+			} => left.is_static() && right.is_static(),
+			Expr::FunctionCall(x) => {
+				// This is not correct as functions like http::get are not 'pure' but this is
+				// replicating previous behavior.
```

Comment:
Not just that, custom functions and `api::invoke` also write (potentially). Not sure how this was done previously though

Reply:
The behavior here is the same as previous. We just do not handle those cases which might result in weird calling behavior. We should handle for each function if it is static or not. However it is also strange to me that some function behave differently if they are considered `static`.

Reply:
This is to do with field definitions:
```
-- these are static, `VALUE` is re-used as `DEFAULT` too
DEFINE FIELD created ON tb_name VALUE time::now();
DEFINE FIELD something ON tb_name VALUE fn::something(123);

-- because we pass a variable to the function here, it's no longer static
-- you would need to add the `DEFAULT` clause manually besides `VALUE`
DEFINE FIELD something ON tb_name VALUE fn::something($param);
```

There are a lot of complexities with field definitions which make them work mostly as easy as they work

Reply:
That is kind of weird, I think `VALUE` should just take the value as it is at the moment of field definition, we have <future>/COMPUTED for if you want the value to change.